README.md
# openneuro-py

A Python client for accessing [OpenNeuro](https://openneuro.org)
datasets.

![openneuro-py in action](openneuro-py.gif)

## Installation

### via conda

```shell
conda install -c conda-forge openneuro-py
```

### via pip

```shell
pip install openneuro-py
```

## Basic usage

### Download an entire dataset

```shell
openneuro-py download --dataset=ds000246
```

### Specify a target directory

To store the downloaded files in a specific directory, use the
`--target_dir` switch. The directory will be created if it doesn't exist
already.

```shell
openneuro-py download --dataset=ds000246 \
                      --target_dir=data/bids
```

### Continue an interrupted download

Interrupted downloads will resume where they left off when you run the command
again.

## Advanced usage
### Exclude a directory from the download

```shell
openneuro-py download --dataset=ds000246 \
                      --exclude=sub-emptyroom
```

### Download only a single file

```shell
openneuro-py download --dataset=ds000246 \
                      --include=sub-0001/meg/sub-0001_coordsystem.json
```

Note that a few essential BIDS files are **always** downloaded in addition.

### Download or exclude multiple files

`--include` and `--exclude` can be passed mulitple times:

```shell
openneuro-py download --dataset=ds000246 \
                      --include=sub-0001/meg/sub-0001_coordsystem.json \
                      --include=sub-0001/meg/sub-0001_acq-LPA_photo.jpg
```


CHANGES.md
# Changelog

## unreleased

- Better handling of server response errors.

## 2022.2.0

- Support latest OpenNeuro API.
- Display suggestions in the exception when `include` contains invalid
  entries.
- Drop list of default excludes. OpenNeuro has fixed server response for the
  respective datasets, so excluding files by default is not necessary anymore.
- Add ability to use an API token to access restricted datasets.

## 2022.1.0

- Fix handling of DOIs that start with `doi:`, as found e.g. in `ds002778`.

## 2021.10.1

- New release for PyPI.

## 2021.10

- Fix unicode terminal detection for Sphinx documentation builds.

## 2021.9

- Add basic support for running Jupyter Notebooks / Jupyter Lab, interactive
  IPython sessions, and in the VS Code interactive Jupyter window.

- Don't crash if the local `dataset_description.json` file is empty when trying
  to resume an aborted download.

- We now by default exclude certain files from the download that are known to
  be invalid for specific datasets. Once the datasets have been fixed on
  OpenNeuro, we will revert these exclusions.

## 2021.8

- Retry downloads if a `ReadError` has occurred.

## 2021.7

- Before resuming a download, check if local and remote datasets actually
  match.

- If neither the API nor the HTTP server provide the file size, don't download
  the file just for size determination. This won't allow us to display progress
  bars anymore in this case, but it will avoid downloading the same file twice.
  Unfortunately, resuming such downloads will be impossible, too: we'll simply
  re-download the files. Note that this only applies to files hosted on
  OpenNeuro.org for which we cannot determine the file size.

- Massively reduce memory footprint (by 2 orders of magnitude).

## 2021.6

- Disable transfer encoding (i.e., compression). This allows for an easier
  check whether a file has been completely downloaded.

## 2021.5

- Ramp up timeouts from 5 to 30 seconds for downloads from `openneuro.org`.
- Drop support for Python 3.6. `openneuro-py` now requires Python 3.7 or newer.
  This change makes development easier.

## 2021.4

- Avoid timeouts that would occur in certain situations.
- Don't stall in situations where a download should be retried.

## 2021.3

- Add Unix path expansion for `openneuro.download()`.
- Support OpenNeuro GraphQL API. New dependencies: `sgqlc` and `requests`.
- Fix crash when not passing any `include`s.

## 2021.2

- Fixes Windows.

## 2021.1

- Fixes for Python <3.8.

## 2021.0

- Improved handling of connection timeouts.

## 2020.7

- Performance improvements.
- Verify file hashes by default whenever possible.

## 2020.6

- Ensure we can operate on Python 3.6.
- Ensure non-Unicode terminals (Windows `cmd`, I'm looking at you!) can be
  used too.

## 2020.5

- Raise an exception if user `--include`s a path that doesn't exist in the
  dataset. Previously, we would silently ignore this issue.
- Optimize checks whether we need to resume a download or not.
- Enable simultaneous downloads (defaults to up to 5 concurrent downloads).
- `openneuro.download()` can now be called from Python scripts directly.

## 2020.4

- Don't rely on OpenNeuro API-reported file sizes anymore, but trust the
  HTTP server. This will do away with error messages, and fix bugs when
  downloading TSV and JSON files, which would sometimes end up being
  incorrectly appended.

## 2020.3.1

- Pin `httpx` requirement to `>=0.15`

## 2020.3

- Switch from `requests` to `httpx` library.
- Add `--verify_size` switch to disable file size sanity check.
- Add auto-retry with backoff on intermittent server errors


openneuro/tests/test_config.py
from pathlib import Path
from unittest import mock

import openneuro
from openneuro.config import init_config, load_config, get_token, Config


def test_config(tmp_path: Path):
    """Test creating and reading the config file."""
    with mock.patch.object(openneuro.config, 'CONFIG_PATH', tmp_path / '.openneuro'):
        assert not openneuro.config.CONFIG_PATH.exists()

        with mock.patch('getpass.getpass', lambda _: 'test'):
            init_config()
        assert openneuro.config.CONFIG_PATH.exists()

        expected_config = Config(endpoint='https://openneuro.org/', apikey='test')
        assert load_config() == expected_config
        assert get_token() == 'test'


openneuro/config.py
from pathlib import Path
import os
import sys
import stat
import json
import getpass
if sys.version_info >= (3, 8):
    from typing import TypedDict
else:
    from typing_extensions import TypedDict

import appdirs
from tqdm.auto import tqdm


CONFIG_DIR = Path(
    appdirs.user_config_dir(appname='openneuro-py', appauthor=False, roaming=True)
)
CONFIG_DIR.mkdir(parents=True, exist_ok=True)
CONFIG_PATH = CONFIG_DIR / 'config.json'
BASE_URL = 'https://openneuro.org/'


class Config(TypedDict):
    endpoint: str
    apikey: str


def init_config() -> None:
    """Initialize a new OpenNeuro configuration file.
    """
    tqdm.write('🙏 Please login to your OpenNeuro account and go to: '
               'My Account → Obtain an API Key')
    api_key = getpass.getpass('OpenNeuro API key (input hidden): ')

    config: Config = {
        'endpoint': BASE_URL,
        'apikey': api_key,
    }

    with open(CONFIG_PATH, 'w', encoding='utf-8') as f:
        json.dump(config, f, indent=2)
    os.chmod(CONFIG_PATH, stat.S_IRUSR | stat.S_IWUSR)


def load_config() -> dict:
    """Load an OpenNeuro configuration file, and return its contents.

    Returns
    -------
    dict
        The configuration options.
    """
    with open(CONFIG_PATH, 'r', encoding='utf-8') as f:
        config = json.load(f)
    return config


def get_token() -> str:
    """Get the OpenNeuro API token if configured with the 'login' command.

    Returns
    -------
    The API token if configured.

    Raises
    ------
    ValueError
        When no token has been configured yet.
    """
    if not CONFIG_PATH.exists():
        raise ValueError(
            'Could not read API token as no openneuro-py configuration '
            'file exists. Run "openneuro login" to generate it.'
        )
    config = load_config()
    if 'apikey' not in config:
        raise ValueError('An openneuro-py configuration file was found, but did not '
                         'contain an "apikey" entry. Run "openneuro login" to '
                         'add such an entry.')
    return config['apikey']


openneuro/openneuro.py
"""
openneuro-py is a leightweight client for accessing OpenNeuro datasets.

Created and maintained by
Richard Höchenberger <richard.hoechenberger@gmail.com>
"""

from pathlib import Path
import requests
from typing import Optional, Tuple

from tqdm import tqdm
import click

from .config import default_base_url


def _get_download_metadata(*,
                           base_url: str,
                           dataset_id: str,
                           tag: Optional[str] = None) -> dict:
    """Retrieve dataset metadata required for the download.
    """
    if tag is None:
        url = f'{base_url}crn/datasets/{dataset_id}/download'
    else:
        url = f'{base_url}crn/datasets/{dataset_id}/snapshots/{tag}/download'

    response = requests.get(url)
    if response.status_code != 200:
        raise RuntimeError(f'Error {response.status_code} when trying to '
                           f'fetch metadata.')

    response_json = response.json()
    return response_json


def _download_files(*,
                    target_dir: Path,
                    files: dict):
    """Download individual files.
    """
    for file in files:
        filename = Path(file['filename'])
        file_size = file['size']
        url = file['urls'][0]

        outfile = target_dir / filename
        outfile.parent.mkdir(parents=True, exist_ok=True)

        response = requests.get(url=url, stream=True)
        if response.status_code != 200:
            raise RuntimeError(f'Error {response.status_code} when trying to '
                               f'download {outfile}.')

        with tqdm.wrapattr(open(outfile, 'wb'),
                           'write',
                           miniters=1,
                           desc=filename.name,
                           dynamic_ncols=True,
                           total=file_size) as f:
            chunk_size = 4096
            for chunk in response.iter_content(chunk_size=chunk_size):
                f.write(chunk)


@click.command()
@click.option('--dataset', required=True, help='The OpenNeuro dataset name.')
@click.option('--tag', help='The tag (version) of the dataset.')
@click.option('--include', multiple=True,
              help='Only include the specified file or directory. Can be '
                   'passed multiple times.')
@click.option('--exclude', multiple=True,
              help='Exclude the specified file or directory. Can be passed '
                   'multiple times.')
def download(*,
             dataset: str,
             tag: Optional[str] = None,
             target_dir: Optional[str] = None,
             include: Optional[Tuple[str]] = None,
             exclude: Optional[Tuple[str]] = None):
    """Download datasets from OpenNeuro.\f

    Parameters
    ----------
    dataset
        The dataset to retrieve, for example ``ds000248``.
    tag
        The tag (revision) of the dataset to retrieve.
    target_dir
        The directory in which to store the downloaded data. If ``None``,
        create a subdirectory with the dataset name in the current working
        directory.
    include
        Files and directories to download. **Only** these files and directories
        will be retrieved.
    exclude
        Files and directories to exclude from downloading.
    """
    if target_dir is None:
        target_dir = Path(dataset)

    include = [] if include is None else include
    exclude = [] if exclude is None else exclude

    metadata = _get_download_metadata(base_url=default_base_url,
                                      dataset_id=dataset,
                                      tag=tag)

    files = []
    for file in metadata['files']:
        filename: str = file['filename']

        # Always include essential BIDS files.
        if filename in ('dataset_description.json',
                        'participants.tsv',
                        'README',
                        'CHANGES'):
            files.append(file)
            continue

        if ((not include or
                any(filename.startswith(i) for i in include)) and
                not any(filename.startswith(e) for e in exclude)):
            files.append(file)

    _download_files(target_dir=target_dir, files=files)


openneuro/download.py
import sys
import hashlib
import asyncio
from pathlib import Path
from typing import Optional, Iterable, Union, Literal

import httpx
from tqdm.asyncio import tqdm
import click
import aiofiles

from . import __version__
from .config import default_base_url


try:
    sys.stdout.reconfigure(encoding='utf-8')
    stdout_unicode = True
except AttributeError:
    # Python 3.6
    stdout_unicode = False


# HTTP server responses that indicate hopefully intermittent errors that
# warrant a retry.
allowed_retry_codes = (408, 500, 502, 503, 504, 522, 524)
allowed_retry_exceptions = (httpx.ConnectTimeout, httpx.ReadTimeout)


def _get_download_metadata(*,
                           base_url: str,
                           dataset_id: str,
                           tag: Optional[str] = None,
                           max_retries: int,
                           retry_backoff: float) -> dict:
    """Retrieve dataset metadata required for the download.
    """
    if tag is None:
        url = f'{base_url}crn/datasets/{dataset_id}/download'
    else:
        url = f'{base_url}crn/datasets/{dataset_id}/snapshots/{tag}/download'

    try:
        response = httpx.get(url)
        request_timed_out = False
    except allowed_retry_exceptions:
        request_timed_out = True

    if request_timed_out and max_retries > 0:
        tqdm.write(f'Request timed out, retrying …')
        asyncio.sleep(retry_backoff)
        max_retries -= 1
        retry_backoff *= 2
        return _get_download_metadata(base_url=base_url, dataset_id=dataset_id,
                                      tag=tag, max_retries=max_retries,
                                      retry_backoff=retry_backoff)
    elif request_timed_out:
        raise RuntimeError('Timeout when trying to fetch metadata.')
    
    if not response.is_error:
        response_json = response.json()
        return response_json
    elif response.status_code in allowed_retry_codes and max_retries > 0:
        tqdm.write(f'Error {response.status_code}, retrying …')
        asyncio.sleep(retry_backoff)
        max_retries -= 1
        retry_backoff *= 2
        return _get_download_metadata(base_url=base_url, dataset_id=dataset_id,
                                      tag=tag, max_retries=max_retries,
                                      retry_backoff=retry_backoff)
    else:
        raise RuntimeError(f'Error {response.status_code} when trying to '
                           f'fetch metadata.')


async def _download_file(*,
                         url: str,
                         api_file_size: int,
                         outfile: Path,
                         verify_hash: bool,
                         verify_size: bool,
                         max_retries: int,
                         retry_backoff: float,
                         semaphore: asyncio.Semaphore) -> None:
    """Download an individual file.
    """
    if outfile.exists():
        local_file_size = outfile.stat().st_size
    else:
        local_file_size = 0

    # Check if we need to resume a download
    # The file sizes provided via the API often do not match the sizes reported
    # by the HTTP server. Rely on the sizes reported by the HTTP server.
    async with semaphore:
        async with httpx.AsyncClient() as client:
            try:
                async with client.stream('HEAD', url=url) as response:
                    headers = response.headers
            except allowed_retry_exceptions:
                if max_retries > 0:
                    await _retry_download(
                        url=url, outfile=outfile,
                        api_file_size=api_file_size,
                        verify_hash=verify_hash, verify_size=verify_size,
                        max_retries=max_retries,
                        retry_backoff=retry_backoff, semaphore=semaphore)
                    return
                else:
                    raise RuntimeError(f'Timeout when trying to download '
                                        f'{outfile}.')

            # Try to get the S3 MD5 hash for the file.
            try:
                remote_file_hash = headers['etag'].strip('"')
                if len(remote_file_hash) != 32:  # It's not an MD5 hash.
                    remote_file_hash = None
            except KeyError:
                remote_file_hash = None

            # Get the Content-Length.
            try:
                remote_file_size = int(response.headers['content-length'])
            except KeyError:
                # The server doesn't always set a Conent-Length header.
                try:
                    async with client.stream('GET', url=url) as response:
                        response_content = await response.aread()
                        remote_file_size = len(response_content)
                except allowed_retry_exceptions:
                    if max_retries > 0:
                        await _retry_download(
                            url=url, outfile=outfile,
                            api_file_size=api_file_size,
                            verify_hash=verify_hash, verify_size=verify_size,
                            max_retries=max_retries,
                            retry_backoff=retry_backoff, semaphore=semaphore)
                        return
                    else:
                        raise RuntimeError(f'Timeout when trying to download '
                                            f'{outfile}.')

    headers = {}
    if outfile.exists() and local_file_size == remote_file_size:
        hash = hashlib.md5()

        if verify_hash and remote_file_hash is not None:
            async with aiofiles.open(outfile, 'rb') as f:
                while True:
                    data = await f.read(65536)
                    if not data:
                        break
                    hash.update(data)

        if (verify_hash and
                remote_file_hash is not None and
                hash.hexdigest() != remote_file_hash):
            desc = f'Re-downloading {outfile.name}: file hash mismatch.'
            mode = 'wb'
            outfile.unlink()
            local_file_size = 0
        else:
            # Download complete, skip.
            desc = f'Skipping {outfile.name}: already downloaded.'
            t = tqdm(iterable=response.aiter_bytes(),
                     desc=desc,
                     initial=local_file_size,
                     total=remote_file_size, unit='B',
                     unit_scale=True,
                     unit_divisor=1024, leave=False)
            t.close()
            return
    elif outfile.exists() and local_file_size < remote_file_size:
        # Download incomplete, resume.
        desc = f'Resuming {outfile.name}'
        headers['Range'] = f'bytes={local_file_size}-'
        mode = 'ab'
    elif outfile.exists():
        # Local file is larger than remote – overwrite.
        desc = f'Re-downloading {outfile.name}: file size mismatch.'
        mode = 'wb'
        outfile.unlink()
        local_file_size = 0
    else:
        # File doesn't exist locally, download entirely.
        desc = outfile.name
        mode = 'wb'

    async with semaphore:
        async with httpx.AsyncClient() as client:
            try:
                async with (
                    client.stream('GET', url=url, headers=headers)
                ) as response:
                    if not response.is_error:
                        pass  # All good!
                    elif (response.status_code in allowed_retry_codes and
                        max_retries > 0):
                        await _retry_download(
                            url=url, outfile=outfile,
                            api_file_size=api_file_size,
                            verify_hash=verify_hash, verify_size=verify_size,
                            max_retries=max_retries,
                            retry_backoff=retry_backoff, semaphore=semaphore)
                        return
                    else:
                        raise RuntimeError(
                            f'Error {response.status_code} when trying '
                            f'to download {outfile} from {url}')

                    await _retrieve_and_write_to_disk(
                        response=response, outfile=outfile, mode=mode,
                        desc=desc,
                        local_file_size=local_file_size,
                        remote_file_size=remote_file_size,
                        remote_file_hash=remote_file_hash,
                        verify_hash=verify_hash, verify_size=verify_size)
            except allowed_retry_exceptions:
                if max_retries > 0:
                    await _retry_download(
                        url=url, outfile=outfile,
                        api_file_size=api_file_size,
                        verify_hash=verify_hash, verify_size=verify_size,
                        max_retries=max_retries,
                        retry_backoff=retry_backoff, semaphore=semaphore)
                    return
                else:
                    raise RuntimeError(f'Timeout when trying to download '
                                       f'{outfile}.')

async def _retry_download(
    *,
    url: str,
    outfile: Path,
    api_file_size: int,
    verify_hash: bool,
    verify_size: bool,
    max_retries: int,
    retry_backoff: float,
    semaphore: asyncio.Semaphore
) -> None:
    tqdm.write('Request timed out, retrying …')
    await asyncio.sleep(retry_backoff)
    max_retries -= 1
    retry_backoff *= 2
    await _download_file(url=url,
                        api_file_size=api_file_size,
                        outfile=outfile,
                        verify_hash=verify_hash,
                        verify_size=verify_size,
                        max_retries=max_retries,
                        retry_backoff=retry_backoff,
                        semaphore=semaphore)


async def _retrieve_and_write_to_disk(
    *,
    response: httpx.Response,
    outfile: Path,
    mode: Literal['ab', 'wb'],
    desc: str,
    local_file_size: int,
    remote_file_size: int,
    remote_file_hash: Optional[str],
    verify_hash: bool,
    verify_size: bool
) -> None:
    hash = hashlib.md5()

    # If we're resuming a download, ensure the already-downloaded
    # parts of the file are fed into the hash function before
    # we continue.
    if verify_hash and local_file_size > 0:
        async with aiofiles.open(outfile, 'rb') as f:
            while True:
                data = await f.read(65536)
                if not data:
                    break
                hash.update(data)

    async with aiofiles.open(outfile, mode=mode) as f:
        with tqdm(desc=desc, initial=local_file_size,
                    total=remote_file_size, unit='B',
                    unit_scale=True, unit_divisor=1024,
                    leave=False) as progress:
            num_bytes_downloaded = response.num_bytes_downloaded

            # TODO Add timeout handling here, too.
            async for chunk in response.aiter_bytes():
                await f.write(chunk)
                progress.update(response.num_bytes_downloaded -
                                num_bytes_downloaded)
                num_bytes_downloaded = (response
                                        .num_bytes_downloaded)
                if verify_hash:
                    hash.update(chunk)

        if verify_hash and remote_file_hash is not None:
            assert hash.hexdigest() == remote_file_hash

        # Check the file was completely downloaded.
        if verify_size:
            await f.flush()
            local_file_size = outfile.stat().st_size
            if not local_file_size == remote_file_size:
                raise RuntimeError(
                    f'Server claimed file size would be {remote_file_size} '
                    f'bytes, but downloaded {local_file_size} byes.')


async def _download_files(*,
                          target_dir: Path,
                          files: Iterable,
                          verify_hash: bool,
                          verify_size: bool,
                          max_retries: int,
                          retry_backoff: float,
                          max_concurrent_downloads: int) -> None:
    """Download files, one by one.
    """
    # Sempahore (counter) to limit maximum number of concurrent download
    # coroutines.
    semaphore = asyncio.Semaphore(max_concurrent_downloads)
    download_tasks = []

    for file in files:
        filename = Path(file['filename'])
        api_file_size = file['size']
        url = file['urls'][0]

        outfile = target_dir / filename
        outfile.parent.mkdir(parents=True, exist_ok=True)
        download_task = _download_file(
            url=url, api_file_size=api_file_size,
            outfile=outfile, verify_hash=verify_hash,
            verify_size=verify_size, max_retries=max_retries,
            retry_backoff=retry_backoff, semaphore=semaphore)
        download_tasks.append(download_task)

    await asyncio.gather(*download_tasks)


def download(*,
             dataset: str,
             tag: Optional[str] = None,
             target_dir: Optional[Union[Path, str]] = None,
             include: Optional[Iterable[str]] = None,
             exclude: Optional[Iterable[str]] = None,
             verify_hash: bool = True,
             verify_size: bool = True,
             max_retries: int = 5,
             max_concurrent_downloads: int = 5) -> None:
    """Download datasets from OpenNeuro.

    Parameters
    ----------
    dataset
        The dataset to retrieve, for example ``ds000248``.
    tag
        The tag (revision) of the dataset to retrieve.
    target_dir
        The directory in which to store the downloaded data. If ``None``,
        create a subdirectory with the dataset name in the current working
        directory.
    include
        Files and directories to download. **Only** these files and directories
        will be retrieved.
    exclude
        Files and directories to exclude from downloading.
    verify_hash
        Whether to calculate and print the SHA256 hash of each downloaded file.
    verify_size
        Whether to check if the downloaded file size matches what the server
        announced.
    max_retries
        Try the specified number of times to download a file before failing.
    max_concurrent_downloads
        The maximum number of downloads to run in parallel.
    """

    msg_problems = 'problems 🤯' if stdout_unicode else 'problems'
    msg_bugs = 'bugs 🪲' if stdout_unicode else 'bugs'
    msg = (f'\n👋 Hello! This is openneuro-py {__version__}. Great to see you! 🤗\n\n'
           f'   👉 Please report {msg_problems} and {msg_bugs} at\n'
           f'      https://github.com/hoechenberger/openneuro-py/issues\n')
    tqdm.write(msg)

    msg = f'Preparing to download {dataset}'
    if stdout_unicode:
        msg = f'🌍 {msg} …'
    else:
        msg += ' ...'
    tqdm.write(msg)

    if target_dir is None:
        target_dir = Path(dataset)
    else:
        target_dir = Path(target_dir)

    include = [] if include is None else list(include)
    exclude = [] if exclude is None else list(exclude)

    retry_backoff = 0.5  # seconds
    metadata = _get_download_metadata(base_url=default_base_url,
                                      dataset_id=dataset,
                                      tag=tag,
                                      max_retries=max_retries,
                                      retry_backoff=retry_backoff)

    files = []
    include_counts = [0] * len(include)  # Keep track of include matches.
    for file in metadata['files']:
        filename: str = file['filename']  # TODO properly define metadata type

        # Always include essential BIDS files.
        if filename in ('dataset_description.json',
                        'participants.tsv',
                        'participants.json',
                        'README',
                        'CHANGES'):
            files.append(file)
            # Keep track of include matches.
            if filename in include:
                include_counts[include.index(filename)] += 1
            continue

        if ((not include or any(filename.startswith(i) for i in include)) and
                not any(filename.startswith(e) for e in exclude)):
            files.append(file)
            # Keep track of include matches.
            for i in include:
                if filename.startswith(i):
                    include_counts[include.index(i)] += 1

    for idx, count in enumerate(include_counts):
        if count == 0:
            raise RuntimeError(f'Could not find paths starting with '
                               f'{include[idx]} in the dataset. Please '
                               f'check your includes.')

    msg = (f'Retrieving up to {len(files)} files '
           f'({max_concurrent_downloads} concurrent downloads).')
    if stdout_unicode:
        msg = f'👉 {msg}'
    tqdm.write(msg)

    # Pre-Python-3.7 compat. Once we drop support for Python 3.6, simply
    # replace this with: asyncio.run(_download_files(...))
    loop = asyncio.get_event_loop()
    loop.run_until_complete(asyncio.wait(
        [_download_files(
            target_dir=target_dir,
            files=files,
            verify_hash=verify_hash,
            verify_size=verify_size,
            max_retries=max_retries,
            retry_backoff=retry_backoff,
            max_concurrent_downloads=max_concurrent_downloads)]
    ))

    msg_finished = f'Finished downloading {dataset}.'
    if stdout_unicode:
        msg_finished = f'✅ {msg_finished}'
    tqdm.write(msg_finished)

    msg_enjoy = 'Please enjoy your brains.'
    if stdout_unicode:
        msg_enjoy = f'\n🧠 {msg_enjoy}'
    else:
        msg_enjoy = f'\n{msg_enjoy}'
    msg_enjoy += '\n'
    tqdm.write(msg_enjoy)


@click.command()
@click.option('--dataset', required=True, help='The OpenNeuro dataset name.')
@click.option('--tag', help='The tag (version) of the dataset.')
@click.option('--target_dir', help='The directory to download to.')
@click.option('--include', multiple=True,
              help='Only include the specified file or directory. Can be '
                   'passed multiple times.')
@click.option('--exclude', multiple=True,
              help='Exclude the specified file or directory. Can be passed '
                   'multiple times.')
@click.option('--verify_hash', type=bool, default=True, show_default=True,
              help='Whether to print the SHA256 hash of each downloaded file.')
@click.option('--verify_size', type=bool, default=True, show_default=True,
              help='Whether to check the downloaded file size matches what '
                   'the server announced.')
@click.option('--max_retries', type=int, default=5, show_default=True,
              help='Try the specified number of times to download a file '
                   'before failing.')
@click.option('--max_concurrent_downloads', type=int, default=5,
              show_default=True,
              help='The maximum number of downloads to run in parallel.')
def download_cli(**kwargs):
    """Download datasets from OpenNeuro."""
    download(**kwargs)
